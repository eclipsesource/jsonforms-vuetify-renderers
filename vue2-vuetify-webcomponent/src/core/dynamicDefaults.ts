import dayjs from 'dayjs';
import customParsingPlugin from 'dayjs/plugin/customParseFormat';
import utcPlugin from 'dayjs/plugin/utc';
import timezonePlugin from 'dayjs/plugin/timezone'; // dependent on utc plugin
import durationPlugin from 'dayjs/plugin/duration';

// required for the custom save formats in the date, time and date-time pickers
dayjs.extend(customParsingPlugin);
dayjs.extend(utcPlugin);
dayjs.extend(timezonePlugin);
dayjs.extend(durationPlugin);

export const searchParams = (args: any) => {
  let paramName: string | undefined = undefined;
  if (args?.param && typeof args.param === 'string') {
    paramName = args?.param;
  }

  const url: URL = new URL(window.location.href);
  const params: URLSearchParams = url.searchParams;

  let result: string | undefined = undefined;
  if (paramName) {
    if (params.has(paramName)) {
      result = params.get(paramName)!;
    } else {
      // try from the hash #
      const hash = url.hash;
      const index = hash.indexOf('?');
      if (index > 0 && index < hash.length - 1) {
        const hashSearchParams = new URLSearchParams(hash.substring(index + 1));
        if (hashSearchParams.has(paramName)) {
          result = hashSearchParams.get(paramName)!;
        }
      }
    }
  }

  return () => result;
};

export const datetimeOffset = (args: any) => {
  const dateTime = nowOffset(args);
  let result: string | undefined = undefined;

  if (dateTime.isValid()) {
    const datetimeLocalFormat = 'YYYY-MM-DDTHH:mm:ss.SSS';

    result = dateTime.local().format(datetimeLocalFormat);
  }

  return () => result;
};

export const timeOffset = (args: any) => {
  const dateTime = nowOffset(args);
  let result: string | undefined = undefined;

  if (dateTime.isValid()) {
    const datetimeLocalFormat = 'HH:mm:ss.SSS';

    result = dateTime.local().format(datetimeLocalFormat);
  }

  return () => result;
};

export const dateOffset = (args: any) => {
  const date = nowOffset(args);
  let result: string | undefined = undefined;

  if (date.isValid()) {
    result = date.local().format('YYYY-MM-DD');
  }

  return () => result;
};

const nowOffset = (args: any) => {
  let duration: string | undefined = undefined;
  if (args?.duration && typeof args.duration === 'string') {
    duration = args?.duration;
  }
  let date: string | Date | undefined = new Date();
  if (args?.date && typeof args.date === 'string') {
    if (args.date !== 'now') {
      // only assign if not the string now
      date = args?.date;
    }
  }
  let operation: 'add' | 'substract' = 'add';
  if (args?.op && args.op === 'substract') {
    operation = 'substract';
  }

  let dateObj = dayjs(date);

  if (dateObj.isValid() && duration) {
    const offset = dayjs.duration(duration);

    dateObj =
      operation == 'add' ? dateObj.add(offset) : dateObj.subtract(offset);
  }

  return dateObj;
};
