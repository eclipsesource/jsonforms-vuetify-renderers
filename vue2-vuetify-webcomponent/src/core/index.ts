export * from './dynamicDefaults';
export * from './i18n';
export * from './json-refs';
export * from './keywords';
export * from './transform';
export * from './types';
export * from './validate';
