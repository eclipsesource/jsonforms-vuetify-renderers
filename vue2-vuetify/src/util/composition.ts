import {
  composePaths,
  computeLabel,
  getFirstPrimitiveProp,
  isDescriptionHidden,
  JsonFormsSubStates,
  Resolve,
} from '@jsonforms/core';
import cloneDeep from 'lodash/cloneDeep';
import merge from 'lodash/merge';
import { useStyles } from '../styles';
import { computed, ComputedRef, inject, ref } from '../vue';
import { Ajv } from 'ajv';

const useControlAppliedOptions = <I extends { control: any }>(input: I) => {
  return computed(() =>
    merge(
      {},
      cloneDeep(input.control.value.config),
      cloneDeep(input.control.value.uischema.options)
    )
  );
};

const useComputedLabel = <I extends { control: any }>(
  input: I,
  appliedOptions: ComputedRef<any>
) => {
  return computed((): string => {
    return computeLabel(
      input.control.value.label,
      input.control.value.required,
      !!appliedOptions.value?.hideRequiredAsterisk
    );
  });
};

/**
 * Adds styles, isFocused, appliedOptions and onChange
 */
export const useVuetifyControl = <
  I extends { control: any; handleChange: any }
>(
  input: I,
  adaptValue: (target: any) => any = (v) => v
) => {
  const appliedOptions = useControlAppliedOptions(input);

  const isFocused = ref(false);
  const onChange = (value: any) => {
    input.handleChange(input.control.value.path, adaptValue(value));
  };

  const persistentHint = (): boolean => {
    return !isDescriptionHidden(
      input.control.value.visible,
      input.control.value.description,
      isFocused.value,
      !!appliedOptions.value?.showUnfocusedDescription
    );
  };

  const computedLabel = useComputedLabel(input, appliedOptions);

  const controlWrapper = computed(() => {
    const { id, description, errors, label, visible, required } =
      input.control.value;
    return { id, description, errors, label, visible, required };
  });

  const styles = useStyles(input.control.value.uischema);

  return {
    ...input,
    styles,
    isFocused,
    appliedOptions,
    controlWrapper,
    onChange,
    persistentHint,
    computedLabel,
  };
};

/**
 * Adds styles and appliedOptions
 */
export const useVuetifyLayout = <I extends { layout: any }>(input: I) => {
  const appliedOptions = computed(() =>
    merge(
      {},
      cloneDeep(input.layout.value.config),
      cloneDeep(input.layout.value.uischema.options)
    )
  );
  return {
    ...input,
    styles: useStyles(input.layout.value.uischema),
    appliedOptions,
  };
};

/**
 * Adds styles, appliedOptions and childUiSchema
 */
export const useVuetifyArrayControl = <I extends { control: any }>(
  input: I
) => {
  const appliedOptions = useControlAppliedOptions(input);

  const computedLabel = useComputedLabel(input, appliedOptions);

  const childLabelForIndex = (index: number) => {
    const childLabelProp =
      input.control.value.uischema.options?.childLabelProp ??
      getFirstPrimitiveProp(input.control.value.schema);
    if (!childLabelProp) {
      return `${index}`;
    }
    const labelValue = Resolve.data(
      input.control.value.data,
      composePaths(`${index}`, childLabelProp)
    );
    if (
      labelValue === undefined ||
      labelValue === null ||
      Number.isNaN(labelValue)
    ) {
      return '';
    }
    return `${labelValue}`;
  };
  return {
    ...input,
    styles: useStyles(input.control.value.uischema),
    appliedOptions,
    childLabelForIndex,
    computedLabel,
  };
};

/**
 * Adds styles and appliedOptions
 */
export const useVuetifyBasicControl = <I extends { control: any }>(
  input: I
) => {
  const appliedOptions = useControlAppliedOptions(input);

  return {
    ...input,
    styles: useStyles(input.control.value.uischema),
    appliedOptions,
  };
};

/**
 * Extracts Ajv from JSON Forms
 */
export const useAjv = () => {
  const jsonforms = inject<JsonFormsSubStates>('jsonforms');

  if (!jsonforms) {
    throw new Error(
      "'jsonforms' couldn't be injected. Are you within JSON Forms?"
    );
  }

  // should always exist
  return jsonforms.core?.ajv as Ajv;
};
