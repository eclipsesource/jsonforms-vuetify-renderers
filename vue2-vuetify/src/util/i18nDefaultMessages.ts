export const i18nDefaultMessages = {
  arraylayout: {
    add: 'Add',
    delete: 'Delete',
    moveUp: 'Move Up',
    moveDown: 'Move Down',
    dialogTitle: 'Delete {{ element }}?',
    dialogText: 'The element will be deleted.',
    dialogConfirm: 'Delete',
    dialogCancel: 'Cancel',
  },
};
