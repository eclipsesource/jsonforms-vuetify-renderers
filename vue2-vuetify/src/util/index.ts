export * from './options';
export * from './composition';
export * from './validator';
export * from './datejs';
export * from './i18nDefaultMessages';
