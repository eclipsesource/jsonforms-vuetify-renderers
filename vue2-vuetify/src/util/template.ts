import templateFn from 'lodash/template';
import merge from 'lodash/merge';
import { TemplateExecutor, TemplateOptions } from 'lodash';

const interpolate = /{{([\s\S]+?)}}/g;

export const template = (
  templateString?: string,
  options?: TemplateOptions
): TemplateExecutor => {
  return templateFn(
    templateString,
    merge({ interpolate: interpolate }, options)
  );
};
