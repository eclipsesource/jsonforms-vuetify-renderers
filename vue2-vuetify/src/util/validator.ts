import { Options } from 'ajv';
import { createAjv as createAjvCore } from '@jsonforms/core';
import Ajv from 'ajv';

export const createAjv = (options?: Options): Ajv => {
  const ajv = createAjvCore(options);
  ajv.addFormat('password', (_) => true);
  return ajv;
};
