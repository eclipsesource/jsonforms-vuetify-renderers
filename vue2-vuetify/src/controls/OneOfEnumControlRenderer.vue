<template>
  <control-wrapper
    v-bind="controlWrapper"
    :styles="styles"
    :isFocused="isFocused"
    :appliedOptions="appliedOptions"
  >
    <v-hover v-slot="{ hover }">
      <v-select
        v-disabled-icon-focus
        :id="control.id + '-input'"
        :class="styles.control.input"
        :disabled="!control.enabled"
        :autofocus="appliedOptions.focus"
        :placeholder="appliedOptions.placeholder"
        :label="computedLabel"
        :hint="control.description"
        :persistent-hint="persistentHint()"
        :required="control.required"
        :error-messages="control.errors"
        :clearable="hover"
        :value="control.data"
        :items="control.options"
        :item-text="(item) => t(item.label, item.label)"
        item-value="value"
        v-bind="vuetifyProps('v-select')"
        @change="onChange"
        @focus="isFocused = true"
        @blur="isFocused = false"
      />
    </v-hover>
  </control-wrapper>
</template>

<script lang="ts">
import {
  ControlElement,
  isOneOfEnumControl,
  JsonFormsRendererRegistryEntry,
  rankWith,
} from '@jsonforms/core';
import {
  rendererProps,
  RendererProps,
  useJsonFormsOneOfEnumControl,
} from '@jsonforms/vue2';
import { defineComponent } from 'vue';
import { VHover, VSelect } from 'vuetify/lib';
import { useTranslator, useVuetifyControl } from '../util';
import { default as ControlWrapper } from './ControlWrapper.vue';
import { DisabledIconFocus } from './directives';

const controlRenderer = defineComponent({
  name: 'oneof-enum-control-renderer',
  components: {
    ControlWrapper,
    VSelect,
    VHover,
  },
  directives: {
    DisabledIconFocus,
  },
  props: {
    ...rendererProps<ControlElement>(),
  },
  setup(props: RendererProps<ControlElement>) {
    const t = useTranslator();

    const control = useVuetifyControl(
      useJsonFormsOneOfEnumControl(props),
      (value) => (value !== null ? value : undefined)
    );
    return { ...control, t };
  },
});

export default controlRenderer;

export const entry: JsonFormsRendererRegistryEntry = {
  renderer: controlRenderer,
  tester: rankWith(5, isOneOfEnumControl),
};
</script>
