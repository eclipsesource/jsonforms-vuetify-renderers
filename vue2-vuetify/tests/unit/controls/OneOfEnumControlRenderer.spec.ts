import { clearAllIds } from '@jsonforms/core';
import { Wrapper } from '@vue/test-utils';
import OneOfEnumControlRenderer, {
  entry as oneOfEnumControlRendererEntry,
} from '../../../src/controls/OneOfEnumControlRenderer.vue';
import { mountJsonForms } from '../util';

describe('OneOfEnumControlRenderer.vue', () => {
  const renderers = [oneOfEnumControlRendererEntry];
  const data = 'a';

  const schema = {
    type: 'string',
    title: 'My OneOf Enum',
    oneOf: [
      { const: 'a', title: 'Foo' },
      { const: 'b', title: 'Bar' },
    ],
  };
  const uischema = {
    type: 'Control',
    scope: '#',
    options: {
      placeholder: 'oneOfEnum placeholder',
    },
  };

  let wrapper: Wrapper<any, Element>;

  beforeEach(() => {
    // clear all ids to guarantee that the snapshots will always be generated with the same ids
    clearAllIds();
    wrapper = mountJsonForms(data, schema, renderers, uischema);
  });

  afterEach(() => {
    wrapper.destroy();
  });

  it('check if child OneOfEnumControlRenderer exists', () => {
    expect(wrapper.getComponent(OneOfEnumControlRenderer));
  });

  it('renders a readonly input', () => {
    expect(
      wrapper.find('input[readonly="readonly"][type="text"]').exists()
    ).toBe(true);
  });

  it('renders title as label', () => {
    expect(wrapper.find('label').text()).toEqual('My OneOf Enum');
  });

  it('emits a data change', async () => {
    const select = wrapper.find('input[type="hidden"]');
    await select.setValue('b');
    expect(wrapper.vm.$data.data).toEqual('b');
  });

  it('should have a placeholder', async () => {
    const input = wrapper.find('input[type="text"]');
    // select the input so the placeholder is generated
    await input.trigger('click');

    const placeholder = input.attributes('placeholder');
    expect(placeholder).toEqual('enum placeholder');
  });

  it('should render component and match snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render component and match snapshot when clicked', async () => {
    const input = wrapper.find('input[type="text"]');
    await input.trigger('click');
    expect(wrapper.html()).toMatchSnapshot();
  });
});
