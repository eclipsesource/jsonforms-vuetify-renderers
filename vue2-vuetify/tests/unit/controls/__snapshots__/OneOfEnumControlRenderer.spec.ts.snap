// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OneOfEnumControlRenderer.vue should render component and match snapshot 1`] = `
<div data-app="true"
     class="v-application v-application--is-ltr theme--light"
     id="app"
>
  <div class="v-application--wrap">
    <div id="#"
         class="control"
         appliedoptions="[object Object]"
         errors
         label="My OneOf Enum"
    >
      <div class="v-input v-input--is-label-active v-input--is-dirty theme--light v-text-field v-select input">
        <div class="v-input__control">
          <div role="button"
               aria-haspopup="listbox"
               aria-expanded="false"
               aria-owns="list-66"
               class="v-input__slot"
          >
            <div class="v-select__slot">
              <label for="#-input"
                     class="v-label v-label--active theme--light"
                     style="left: 0px; position: absolute;"
              >
                My OneOf Enum
              </label>
              <div class="v-select__selections">
                <div class="v-select__selection v-select__selection--comma">
                  Foo
                </div>
                <input id="#-input"
                       readonly="readonly"
                       type="text"
                       aria-readonly="false"
                       autocomplete="off"
                >
              </div>
              <div class="v-input__append-inner">
                <div class="v-input__icon v-input__icon--append">
                  <i aria-hidden="true"
                     class="v-icon notranslate mdi mdi-menu-down theme--light"
                  >
                  </i>
                </div>
              </div>
              <input type="hidden"
                     value="a"
              >
            </div>
            <div class="v-menu">
            </div>
          </div>
          <div class="v-text-field__details">
            <div class="v-messages theme--light">
              <transition-group-stub tag="div"
                                     name="message-transition"
                                     class="v-messages__wrapper"
              >
              </transition-group-stub>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`OneOfEnumControlRenderer.vue should render component and match snapshot when clicked 1`] = `
<div data-app="true"
     class="v-application v-application--is-ltr theme--light"
     id="app"
>
  <div class="v-application--wrap">
    <div id="#"
         class="control"
         appliedoptions="[object Object]"
         errors
         label="My OneOf Enum"
         isfocused="true"
    >
      <div class="v-input v-input--is-label-active v-input--is-dirty v-input--is-focused theme--light v-text-field v-select v-select--is-menu-active primary--text input">
        <div class="v-input__control">
          <div role="button"
               aria-haspopup="listbox"
               aria-expanded="true"
               aria-owns="list-79"
               class="v-input__slot"
          >
            <div class="v-select__slot">
              <label for="#-input"
                     class="v-label v-label--active theme--light primary--text"
                     style="left: 0px; position: absolute;"
              >
                My OneOf Enum
              </label>
              <div class="v-select__selections">
                <div class="v-select__selection v-select__selection--comma">
                  Foo
                </div>
                <input id="#-input"
                       readonly="readonly"
                       type="text"
                       aria-readonly="false"
                       autocomplete="off"
                >
              </div>
              <div class="v-input__append-inner">
                <div class="v-input__icon v-input__icon--append">
                  <i aria-hidden="true"
                     class="v-icon notranslate mdi mdi-menu-down theme--light primary--text"
                  >
                  </i>
                </div>
              </div>
              <input type="hidden"
                     value="a"
              >
            </div>
            <div class="v-menu">
              <transition-stub name="v-menu-transition">
              </transition-stub>
            </div>
          </div>
          <div class="v-text-field__details">
            <div class="v-messages theme--light primary--text">
              <transition-group-stub tag="div"
                                     name="message-transition"
                                     class="v-messages__wrapper"
              >
              </transition-group-stub>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div class="v-menu__content theme--light menuable__content__active "
       style="max-height: 304px; min-width: 0px; max-width: auto; top: 12px; left: 0px; transform-origin: top left; z-index: 8; display: none;"
  >
    <div role="listbox"
         tabindex="-1"
         class="v-list v-select-list v-sheet theme--light theme--light"
         id="list-79"
    >
      <div tabindex="0"
           aria-selected="true"
           id="list-item-86-0"
           role="option"
           class="v-list-item primary--text v-list-item--active v-list-item--link theme--light"
      >
        <div class="v-list-item__content">
          <div class="v-list-item__title">
            Foo
          </div>
        </div>
      </div>
      <div tabindex="0"
           aria-selected="false"
           id="list-item-86-1"
           role="option"
           class="v-list-item v-list-item--link theme--light"
      >
        <div class="v-list-item__content">
          <div class="v-list-item__title">
            Bar
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
